src/game/editor/editor2.cpp

```diff
@@ -2750,6 +2757,265 @@ void CEditor2::RenderPopupBrushPalette()
 	RenderBrush(m_UiMousePos);
 }
 
+void CEditor2::InvokePopupFileSelect(const char *pInitialPath, bool (*pfnCallback)(const char *pFilename, void *pContext), void *pContext)
+{
+	m_UiFileSelectState.m_Selected = -1;
+	m_UiFileSelectState.m_pContext = pContext;
+	m_UiFileSelectState.m_pfnMapSelectCB = pfnCallback;
+	str_copy(m_UiFileSelectState.m_aPath, pInitialPath, sizeof(m_UiFileSelectState.m_aPath));
+	m_UiFileSelectState.m_ppListBoxEntries = 0;
+
+	m_UiFileSelectState.PopulateFileList(Storage(), IStorage::TYPE_SAVE);
+
+	m_UiCurrentPopupID = POPUP_FILE_SELECT;
+}
+
+int CEditor2::CUIFileSelect::EditorListdirCallback(const char *pName, int IsDir, int StorageType, void *pUser)
+{
+	CUIFileSelect *pState = (CUIFileSelect *)pUser;
+
+	if(!str_comp(pName, "."))
+		return 0;
+
+	CEditor2::CFileListItem Item;
+	str_copy(Item.m_aFilename, pName, sizeof(Item.m_aFilename));
+	if(IsDir)
+		str_format(Item.m_aName, sizeof(Item.m_aName), "%s/", pName);
+	else
+		str_copy(Item.m_aName, pName, sizeof(Item.m_aName));
+
+	Item.m_IsDir = IsDir != 0;
+	Item.m_IsLink = false;
+	Item.m_StorageType = StorageType;
+	pState->m_aFileList.add(Item);
+
+	return 0;
+}
+
+void CEditor2::CUIFileSelect::PopulateFileList(IStorage *pStorage, int StorageType)
+{
+	m_aFileList.clear();
+	pStorage->ListDirectory(StorageType, m_aPath, EditorListdirCallback, this);
+	GenerateListBoxEntries();
+
+	pStorage->GetCompletePath(IStorage::TYPE_SAVE, m_aPath, m_aCompletePath, sizeof(m_aCompletePath));
+}
+
+void CEditor2::CUIFileSelect::GenerateListBoxEntries()
+{
+	// An array<const char *> might help get allocations down
+	if(m_ppListBoxEntries)
+		free(m_ppListBoxEntries);
+
+	m_ppListBoxEntries = (const char **)mem_alloc(m_aFileList.size() * 3 * sizeof(*m_ppListBoxEntries), 0);
+	for(int i = 0; i < m_aFileList.size(); i++)
+	{
+		m_ppListBoxEntries[i * 3] = m_aFileList[i].m_aName;
+		m_ppListBoxEntries[i * 3 + 1] = "D1";
+		m_ppListBoxEntries[i * 3 + 2] = "D2";
+	}
+}
+
+void CEditor2::RenderPopupFileSelect()
+{
+	const float Padding = 20.0f;
+	const float Scale = 2.0f/3.0f;
+	const float FontSize = 7.0f;
```

Comment:
We never use a font this small in Teeworlds, so this softly breaks some of the smallest resolutions. Would be possible to increase it a bit?

Reply:
Sure but it'd look awkward, is there a better way to fix this?

Reply:
If you look at the demos file browser, it uses larger fonts/boxes:

![screenshot_2020-03-25_00-30-35](https://user-images.githubusercontent.com/355114/77486422-02a3db80-6e30-11ea-802b-0bc0add74cb0.png)

This editor2 interface looks good, but only on large resolutions. Font 8, even 9 doesn't look awkward to me:

![image](https://user-images.githubusercontent.com/355114/77486659-a3929680-6e30-11ea-989e-2a0526fa7265.png)

though the new folder button is quite tight. Even without the upscale, locales tend to be longer than english (if you're german, you gotta know what I mean ;) )

Reply:
As resolutions get smaller it get harder to lay things out. I dont ever see how that new folder text fits in that button. We might need to either have a button with a new folder icon, or employ some of the modern responsive design tricks. E.g. move the favourites out of the way on smaller resolutions.

Reply:
I don't get why we should support such small screen sizes in editor. Working with such a small window is nothing you really want if you are serious about map making.

Reply:
Another idea would be to have a fallback icon. If the text doesnt fit, use an icon instead.

Reply:
we should use as much icons as possible and add tooltips, like every other pro software is doing.

Reply:
@Learath2 The `New folder` button is a problem anyway because of locales. `Nouveau dossier` doesn't fit in there, for instance. All buttons in Teeworlds have some spare width because of that. You could fit it below Favorites?

I think we should just use more space. It's unlikely we end up with the three columns anyway.

The original issue I was raising was the record-low font size though, not the button spacing, that can be adjusted later :)

Reply:
we can also add a new 5th button in the top for "new folder", with just an icon.

Reply:
Well the back and forward buttons arent really too useful, we could maybe get rid of those?

Reply:
i think we have not space problem on top, I added them because we are used to them in normal filebrowsers.

Reply:
![image](https://user-images.githubusercontent.com/490500/77532611-40434b80-6ea6-11ea-86e6-1b6ba700475f.png)

I think this works better anyway

Reply:
Yes looks good.
Icons:
![editor_icons](https://user-images.githubusercontent.com/329471/77533006-58ab6a00-6e96-11ea-96ee-ddc8e99aeb7e.png)

Reply:
The form does not matter for now, all that matter is that it is functional. We'll have this conversation again much later on, when UI design comes into consideration. The code looks good for now @Learath2 👍 .